api/sms_provider/custom.go

```diff
@@ -0,0 +1,57 @@
+package sms_provider
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/netlify/gotrue/conf"
+)
+
+type CustomProvider struct {
+	Config *conf.CustomProviderConfiguration
+}
+
+// Creates a SmsProvider with the custom Config
+func NewCustomProvider(config conf.CustomProviderConfiguration) (SmsProvider, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &CustomProvider{
+		Config: &config,
+	}, nil
+}
+
+// Send an SMS containing the OTP with custom URL
+func (t *CustomProvider) SendSms(phone string, message string) error {
+	body, err := json.Marshal(map[string]string{
+		"recipient": phone,
+		"body":      message,
+		"sender":    t.Config.Sender,
+	})
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{Timeout: defaultTimeout}
+	r, err := http.NewRequest("POST", t.Config.Url, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+	r.Header.Add("Content-Type", "application/json")
+	if len(t.Config.BearerToken) > 0 {
+		r.Header.Add("Authorization", "Bearer "+t.Config.BearerToken)
+	}
+	res, err := client.Do(r)
+	if err != nil {
+		return err
+	}
+
+	if res.StatusCode/100 != 2 {
+		return fmt.Errorf("Unexpected response while calling the SMS gateway: %v", res.StatusCode)
+	}
+
+	return nil
```

Comment:
I would suggest using a more generic approach here as mentioned in one of the previous comments. Instead of specifying an `Authorization` header specifically, it is better and more flexible in my opinion to set any headers that you may need -- like API keys or other special values.

Furthermore, I don't see the need for the `"sender"` field. You only need to post the message and recipient phone number.

Finally, please handle HTTP errors and redirects correctly. For example, if the URL returns a 301 response, what happens? What's the timeout?

Reply:
I just followed what Auth0 is already doing and I believe it covers most use cases. It's not a matter of personal opinion; It's best we KISS.

Reply:
It returns an error for any non-2xx response similar to the twilio provider. Also it uses the default timeout similar to all external providers.